Enable and use the automatic limit for array-as-uninterpreted-function (duplicated, fixes 6130)

TGWDB opened the pull request:
This is a recreation of PR #2108 that should fix Issue #6130 . This has been created as a PR to run CI after rebasing on develop (as of this morning).

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

TGWDB committed:
Enable and use the automatic limit for array-as-uninterpreted-function

This is a duplicate of PR #2108. This is here for alternative author
management.

Fixes #6130

TGWDB committed:
Enable and use the automatic limit for array-as-uninterpreted-function

This is a duplicate of PR #2108. This is here for alternative author
management.

Fixes #6130

TGWDB committed:
Merge branch 'issue6130-auto-array' of github.com:TGWDB/cbmc into issue6130-auto-array

TGWDB commented:
@martin-cs Any chance of approval, this is the same as #2108 that you approved (just rebased and passing newer CI).

martin-cs commented on src/solvers/flattening/boolbv_width.cpp:
```diff
@@ -136,13 +135,7 @@ const boolbv_widtht::entryt &boolbv_widtht::get_entry(const typet &type) const
-      if(total>(1<<30)) // realistic limit
-        throw analysis_exceptiont("array too large for flattening");
-
-      if(total < 0)
```
This case does not seem to be handled any more.  Maybe it could become an invariant?

TGWDB replied:
I did some checking and this error handling should be left in since it can trigger an exception if not handled. I've pushed a fix here.

martin-cs reviewed (approved):
As I said in #2108 , I am OK with this but there is some kind of bug in equality of unbounded arrays which I have yet to reliably trigger via C or locate. I figure that this might expose it so it would be good to have a few tests of what happens when an array is over `MAX_FLATTENED_ARRAY_SIZE`.

TGWDB committed:
Error handling for very large size arrays and vectors

This reinstates and improves the error handling for very large
arrays and vectors with the limit of 1 << 30 (from old code).
This prevents exceptions in the cast which can occur for extremely
large values.

TGWDB committed:
Add regression test for size limit check

This adds a regression test to check that the size limit for
arrays is caught and cleanly handled (with appropriate error).

TGWDB committed:
Revert "Add regression test for size limit check"

This reverts commit bc0256555c53082705fb6e9b111d47471cb194ee.
